packages/next/src/client/components/navigation.ts

```diff
@@ -167,11 +169,25 @@ function getSelectedParams(
 export function useParams(): Params {
   clientHookInServerComponentError('useParams')
   const globalLayoutRouterContext = useContext(GlobalLayoutRouterContext)
-  if (!globalLayoutRouterContext) {
-    // This only happens in `pages`. Type is overwritten in navigation.d.ts
-    return null!
+  const pagesRouter = useContext(RouterContext)
+
+  // When it's under app router
+  if (globalLayoutRouterContext) {
+    return getSelectedParams(globalLayoutRouterContext.tree)
+  }
+
+  // When it's under pages router,
+  if (pagesRouter) {
```

Comment:
Maybe `if (pagesRouter?.isReady)`?

Reply:
`isReady` of server router is `false`, if we only extract when `isReady` is `true`, you would see a flicking since the SSR content and client content could be mismatch